### DISCUSSION TITLE
How to run Diesel migrations with rocket_db_pools and async?

### DISCUSSION THREAD

**Discussion by blackghost1987** [rwf2/Rocket#2640] [Questions] [answered]
I started to upgrade my project to rc.4 (which went fine by the way), and I just realized there's now support for Async Postgresql with Diesel, so I tried to refactor my code to use that as well. I had some issues figuring out that I need to rewrite my imports to use rocket_db_pools::diesel instead of diesel to get the async variants, but after fixing that, it started to work.

The only issue I have left is how to run the Diesel migrations automatically on startup?

I found the example using fairings to do it, but I cannot get it to work properly with async Diesel connection pool.

Here's what I got:
```rust
use diesel_migrations::{FileBasedMigrations, MigrationHarness};
use rocket::{fairing, Build, Rocket};
use rocket_db_pools::diesel::dsl::*;
use rocket_db_pools::diesel::prelude::*;
use rocket_db_pools::diesel::RunQueryDsl;
use rocket_db_pools::{Connection, Database};

pub type ConnPool = rocket_db_pools::diesel::PgPool;

#[derive(Database)]
#[database("my_db")]
pub struct Db(ConnPool);

pub async fn run_migrations(rocket: Rocket<Build>) -> fairing::Result {
    let migrations = FileBasedMigrations::find_migrations_directory().expect("migrations directory not found");
    match Db::fetch(&rocket) {
        Some(mut db) => match db.0.run_migrations(migrations) {
            Ok(_) => {
                info!("Migrations executed successfully");
                Ok(rocket)
            },
            Err(e) => {
                error!("Migrations failed! Error: {:?}", e);
                Err(rocket)
            },
        },
        None => Err(rocket),
    }
}

...

fn build_rocket() -> Rocket<Build> {
    rocket::build()
        .attach(Db::init())
        .attach(AdHoc::try_on_ignite("Diesel Migrations", run_migrations))
        .mount("/", routes![index, favicon])
}

```

I tried both `match db.run_migrations` and `match db.0.run_migrations`, both giving me trait bound issues, like the following:

```
error[E0599]: the method `run_migrations` exists for reference `&Db`, but its trait bounds were not satisfied
   --> src\database.rs:42:34
    |
27  |   pub struct Db(ConnPool);
    |   -------------
    |   |
    |   doesn't satisfy `<Db as diesel::Connection>::Backend = _`
    |   doesn't satisfy `Db: MigrationConnection`
    |   doesn't satisfy `Db: MigrationHarness<_>`
    |   doesn't satisfy `Db: diesel::Connection`
...
42  |           Some(mut db) => match db.run_migrations(migrations) {
    |                                    ^^^^^^^^^^^^^^ method cannot be called on `&Db` due to unsatisfied trait bounds
    |
   ::: C:\Users\blackghost\.cargo\registry\src\index.crates.io-6f17d22bba15001f\diesel-2.1.3\src\query_builder\delete_statement\mod.rs:25:1
    |
25  |   pub struct DeleteStatement<T: QuerySource, U, Ret = NoReturningClause> {
    |   ----------------------------------------------------------------------
    |   |
    |   doesn't satisfy `_: ExecuteDsl<&Db, _>`
    |   doesn't satisfy `_: ExecuteDsl<Db, _>`
    |   doesn't satisfy `_: ExecuteDsl<Pool<AsyncDieselConnectionManager<AsyncPgConnection>>, _>`
```

and

```
error[E0599]: the method `run_migrations` exists for struct `Pool<AsyncDieselConnectionManager<AsyncPgConnection>>`, but its trait bounds were not satisfied

43  |           Some(mut db) => match db.0.run_migrations(migrations) {
    |                                      ^^^^^^^^^^^^^^ method cannot be called due to unsatisfied trait bounds
    |
   ::: C:\Users\blackghost\.cargo\registry\src\index.crates.io-6f17d22bba15001f\diesel-2.1.3\src\query_builder\delete_statement\mod.rs:25:1
    |
25  |   pub struct DeleteStatement<T: QuerySource, U, Ret = NoReturningClause> {
    |   ---------------------------------------------------------------------- doesn't satisfy `_: ExecuteDsl<Pool<AsyncDieselConnectionManager<AsyncPgConnection>>, _>`
```

Any ideas how to get it working? What I find strange is that the run_migrations function is actually sync, is that normal? Or am I trying to call the wrong function here and there should be an async equivalent?

**Comment by blackghost1987** [comment]
Ahh I think I found the issue: `diesel-async` only started to support `diesel_migrations` since 0.4.0, when they added a type conversion from `diesel-async::AsyncConnection` to `diesel::Connection` which would hopefully make this work. Unfortunately `rocket_db_pools` rc.4 is using version ^0.3.1 of `diesel-async`.

(also I realized I should have used `run_pending_migrations` in my code instead of `run_migrations`)

I'll take a look at how straightforward the upgrade is, and open a PR if possible.

**Comment by blackghost1987** [comment]
`diesel-async` has been updated to 0.4.1 and the latest Rocket has been released with version 0.5.0. Unfortunately updating everything didn't help with my issue.

I tried replicating the example in `diesel-async` here:
https://github.com/weiznich/diesel_async/blob/main/examples/postgres/run-pending-migrations-with-rustls/src/main.rs
Adding the AsyncConnectionWrapper didn't help either, I still have a type mismatch.

So I still don't have any idea how to run pending migrations with async Diesel, any help is appreciated!

**Reply by SergioBenitez** [reply, marked as answer]
Check out #2648, where @ThouCheese uses diesel-async with migrations.

**Comment by blackghost1987** [comment]
Sorry for leaving this open for so long, the solution in the linked PR works (although it's still not merged)
